core/src/main/scala/org/apache/spark/BarrierCoordinator.scala

```diff
@@ -134,11 +138,8 @@ private[spark] class BarrierCoordinator(
 
     // Cancel the current active TimerTask and release resources.
     private def cancelTimerTask(): Unit = {
-      if (timerTask != null) {
-        timerTask.cancel()
-        timer.purge()
-        timerTask = null
-      }
+      timerFutures.asScala.foreach(_.cancel(false))
```

Comment:
What if a new future is added in between these two lines ?

I think the correct way to cancel and remove the futures is to iterate over the futures in a block enclosed within `timerFutures.synchronized`

Reply:
A new future will not come because when a stage is of type Barrier all the tasks are started together and the stage waits for all the tasks to complete before proceeding. Therefore, there will not be any future that will be added in between.  Secondly, `cancelTimerTask` is called from within a `synchronized` block in the 2 places it is being called from

Reply:
Thanks for your comment.
I think merging the removal of futures with cancellation would make the code amenable to future code changes.